controllers/installation_controller.go

```diff
@@ -0,0 +1,179 @@
+}
+
+// NodeHasChanged returns true if the node configuration has changed when compared to
+// the node information we keep in the installation status. Returns a bool indicating
+// if a change was detected and a bool indicating if the node is new (not seen yet).
+func (r *InstallationReconciler) NodeHasChanged(in *v2alpha1.Installation, ev metrics.NodeEvent) (bool, bool, error) {
+	for _, nodeStatus := range in.Status.NodesStatus {
+		if nodeStatus.Name != ev.NodeName {
+			continue
+		}
+		eventHash, err := ev.Hash()
+		if err != nil {
+			return false, false, err
+		}
+		return nodeStatus.Hash != eventHash, false, nil
+	}
+	return true, true, nil
+}
+
+// UpdateNodeStatus updates the node status in the Installation object status.
+func (r *InstallationReconciler) UpdateNodeStatus(in *v2alpha1.Installation, ev metrics.NodeEvent) error {
+	hash, err := ev.Hash()
+	if err != nil {
+		return err
+	}
+	for i, nodeStatus := range in.Status.NodesStatus {
+		if nodeStatus.Name != ev.NodeName {
+			continue
+		}
+		in.Status.NodesStatus[i].Hash = hash
+		return nil
+	}
+	in.Status.NodesStatus = append(in.Status.NodesStatus, v2alpha1.NodeStatus{Name: ev.NodeName, Hash: hash})
+	return nil
+}
+
+// ReconcileInstallation is the function that actually reconciles the Installation object.
+// Metrics events (call back home) are generated here.
+func (r *InstallationReconciler) ReconcileInstallation(ctx context.Context, in *v2alpha1.Installation) error {
+	var nodes corev1.NodeList
+	if err := r.List(ctx, &nodes); err != nil {
+		return fmt.Errorf("failed to list nodes: %w", err)
+	}
+	seen := map[string]bool{}
+	needsUpdate := false
+	for _, node := range nodes.Items {
+		seen[node.Name] = true
+		event := metrics.NodeEventFromNode(in.Spec.ClusterID, node)
+		changed, isnew, err := r.NodeHasChanged(in, event)
+		if err != nil {
+			return fmt.Errorf("failed to check if node has changed: %w", err)
+		} else if !changed {
+			continue
+		}
+		needsUpdate = true
+		if err := r.UpdateNodeStatus(in, event); err != nil {
+			return fmt.Errorf("failed to update node status: %w", err)
+		}
+		if isnew {
+			if err := metrics.NotifyNodeAdded(ctx, in.Spec.MetricsBaseURL, event); err != nil {
+				return fmt.Errorf("failed to notify node added: %w", err)
+			}
+			continue
+		}
+		if err := metrics.NotifyNodeUpdated(ctx, in.Spec.MetricsBaseURL, event); err != nil {
+			return fmt.Errorf("failed to notify node updated: %w", err)
+		}
+	}
+	trimmed := []v2alpha1.NodeStatus{}
+	for _, nodeStatus := range in.Status.NodesStatus {
+		if _, ok := seen[nodeStatus.Name]; ok {
+			trimmed = append(trimmed, nodeStatus)
+			continue
+		}
+		needsUpdate = true
+		rmevent := metrics.NodeRemovedEvent{ClusterID: in.Spec.ClusterID, NodeName: nodeStatus.Name}
+		if err := metrics.NotifyNodeRemoved(ctx, in.Spec.MetricsBaseURL, rmevent); err != nil {
+			return fmt.Errorf("failed to notify node removed: %w", err)
+		}
+	}
+	log := ctrl.LoggerFrom(ctx)
+	in.Status.NodesStatus = trimmed
```

Comment:
I have a feeling we should sort this (by node name, for example)

Reply:
Why ?

Reply:
just to make sure that it's stable - order changing _shouldn't_ cause issues, but we should avoid it anyways

Reply:
Yeah, agreed. Let me work on it.

Reply:
Done.

Reply:
don'tkillme.jpg

https://pkg.go.dev/sort#SliceStable

```suggestion
        sort.SliceStable(trimmed, func(i, j int) bool { return trimmed[i].Name < trimmed[j].Name })
	in.Status.NodesStatus = trimmed
```

instead of making the custom type and everything